### ISSUE TITLE
Reusing or sharing tasks

### ISSUE THREAD

**Issue by pwm1234-sri** [nightroman/Invoke-Build#195]
I have started learning Invoke-Build, and I think it will be a great help to me. 
One question that I have is how do I reuse code or tasks? 
How can I share an Invoke-Build task across multiple projects? 
Or is the right way to put the reusable logic into a function in a script or module that is imported in a typical powershell way?

**Comment by nightroman** [comment]
Here is the most straightforward and idiomatic way.

(1) Compose your shared tasks in a separate script, example: `Shared.tasks.ps1`, https://github.com/nightroman/Invoke-Build/blob/master/Tests/Shared.tasks.ps1

(2) Import tasks into a build script. Two ways.

(2.1) If the task script contains just tasks with no script scope variables
(e.g. `SharedTask1`, `SharedTask2`) then simply invoke the task script from
a build script:

https://github.com/nightroman/Invoke-Build/blob/01517872acd7ad945ab46c5e22b65ee3df5a4056/Tests/.build.ps1#L40-L42

(2.2) If shared tasks use script scope variables (e.g. `Sleep` uses
`$SleepMilliseconds`) then dot-source the task script:

```pwsh
. .\Shared.tasks.ps1 5000
```

(Granted, this is not a typical example, the variable is also a parameter.)

One way or another, by invoking the task script from a build script, you get
the shared tasks added (note that `task` is just an alias of `Add-BuildTask`)
to other tasks defined (added) in the build script.

**Another existing example**

Task script: https://github.com/nightroman/FarNet/blob/master/FarNet/FarNet.tasks.ps1

This task script is imported into (invoked from) several build scripts:

- https://github.com/nightroman/FarNet/blob/master/FarNet/FarNet.Tools/.build.ps1
- https://github.com/nightroman/FarNet/blob/master/FarNet/FarNet.Works.Config/.build.ps1
- a few more projects

Note that in this example actual build scripts are almost "empty", they all
basically consists of the imported tasks. But they use different environments
(script variables).

**Comment by nightroman** [comment]
> Or is the right way to put the reusable logic into a function in a script or module that is imported in a typical powershell way?

This is another absolutely right way. You share the usual PowerShell functions / scripts / modules that do some job. Then you call them from build script tasks. Tasks are just thin wrappers of that calls.

Which way to use? You decide which is better in a particular case. I use both and kind of prefer imported task files when tasks do not require (many) parameters / script variables.

With many parameters / variables, especially different for different tasks, the shared tasks way may be less attractive, more messy. Then I use shared functions / script / modules and call them from thin tasks.

**Comment by pwm1234-sri** [comment]
Thank you for your detailed answer! 
At the moment as I am just getting started, my needs are rather simple, so I will use imported tasks. 
If my needs grow (and my powershell scripting skills improve) I will move to a module.

**Comment by nightroman** [comment]
Oh, I completely forgot :) There is the whole article about this, with examples, including shared tasks in modules.
https://github.com/nightroman/Invoke-Build/tree/master/Tasks/Import

**Comment by pwm1234-sri** [comment]
Thank you! Now that I am understanding invoke-build a little better, I believe there are a number of goodies in the Tasks folder that I should study. Thanks again!!!
